### PR TITLE
⚠️ in-place propagation from MS to InfraMachine and BootstrapConfig

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:

This PR adds in-place propagation support for labels and annotation from MachineSet to InfraMachine and BootstrapConfig.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

Part of #7731

### CODE REVIEW

File: controllers/external/util.go
Comment (sbueringer) [inline controllers/external/util.go:89]:
```suggestion
	// When SSA is set to true, SSAPatchOptions cannot be empty.
```
nit

File: controllers/external/util.go
Comment (sbueringer) [inline controllers/external/util.go:98]: I think it would be cleaner to add a new ApplyFromTemplate.

Preserves the current CreateFromTemplate as is and makes it less awkward to use CreateFromTemplate for updates/applies

File: internal/controllers/machineset/machineset_controller.go
Comment (sbueringer) [inline internal/controllers/machineset/machineset_controller.go:434]: I think there should be a way to re-use ApplyFromTemplate / CreateFromTemplate here. The only diff should be a few Names/UIDs? We can probably just hand over the old object and take name and uid from there if it exists? (we did something similar in the topology reconciler)

File: internal/controllers/machineset/machineset_controller.go
Comment (sbueringer) [inline internal/controllers/machineset/machineset_controller.go:425]: Good question. We also never did set it in the topology reconciler.

I think given that we always want to write our intent independent of what the object on the server-side looks like it should be fine.

File: internal/controllers/machineset/machineset_controller.go
Comment (sbueringer) [reply internal/controllers/machineset/machineset_controller.go:434]: I think it would be simpler this way and we avoid divergense between create and apply later.

**Review by sbueringer — COMMENTED** [review]
Overall looks good to me

File: internal/controllers/machinedeployment/machinedeployment_controller.go
Comment (ykakarap) [inline internal/controllers/machinedeployment/machinedeployment_controller.go:413]: This will be dropped in favor of the ssa util package that will be introduced in #7921.

File: internal/controllers/machineset/machineset_controller.go
Comment (sbueringer) [inline internal/controllers/machineset/machineset_controller.go:657]:
```suggestion
	machineLabels := map[string]string{}
```
nit: just because I saw it in Intellij

**Review by sbueringer — COMMENTED** [review]
Overall direction sounds good to me.

I would do one final round of nitpickery once we merged the underlying PR

File: controllers/external/util.go
Comment (fabriziopandini) [inline controllers/external/util.go:214]: If I got this right we are introducing this func in order to not call two times Get, not sure it is worth the effort.
I will keep things simple and avoid introducing this layer on top of GenerateTemplate

File: internal/controllers/machineset/machineset_controller.go
Comment (fabriziopandini) [inline internal/controllers/machineset/machineset_controller.go:617]: I struggle a little bit with the semantics of this func, but we can iterate back as soon as we reach an agreement on 7291

**Review by fabriziopandini — COMMENTED** [review]
From a first look the approach is ok, and with some little refactoring we can make the code simpler to read and to unit test

File: controllers/external/util.go
Comment (sbueringer) [reply controllers/external/util.go:214]: +1

File: internal/controllers/machineset/machineset_controller.go
Comment (sbueringer) [inline internal/controllers/machineset/machineset_controller.go:662]:
```suggestion
	updatedObject := &unstructured.Unstructured{}
```

File: internal/controllers/machineset/machineset_controller.go
Comment (sbueringer) [inline internal/controllers/machineset/machineset_controller.go:647]: I know it's trivial at the moment but let's also extract a machineAnnotationsFromMachineSet func to ensure we use the same code in both cases

File: internal/controllers/machineset/machineset_controller_test.go
Comment (sbueringer) [inline internal/controllers/machineset/machineset_controller_test.go:1051]: nit: Let's also add a few fields that should stay the same (same for BootstrapConfig)

File: internal/controllers/machineset/machineset_controller_test.go
Comment (sbueringer) [inline internal/controllers/machineset/machineset_controller_test.go:1195]: Let's extend this validation once we changed the prod code to verify that the ssa manager owns labels/annotations and "manager" some other spec field that we added

prior art: https://github.com/kubernetes-sigs/cluster-api/blob/4d3b1817e3ee53e34cef21575bb29ec2575b4744/internal/controllers/topology/cluster/structuredmerge/serversidepathhelper_test.go#L696

Let's also check that the spec field stayed the same

(same for bootstrapconfig of course)

File: internal/controllers/machineset/machineset_controller.go
Comment (sbueringer) [inline internal/controllers/machineset/machineset_controller.go:411]: Let's discuss. I would prefer cleaning up only labels & annotations

Prior art ~ https://github.com/kubernetes-sigs/cluster-api/blob/4d3b1817e3ee53e34cef21575bb29ec2575b4744/internal/controllers/topology/cluster/structuredmerge/dryrun.go#L195-L202

Comment (sbueringer) [reply internal/controllers/machineset/machineset_controller.go:411]: I would move filterintent.go into the ssa util package

File: internal/util/ssa/matchers.go
Comment (sbueringer) [inline internal/util/ssa/matchers.go:67]: Nice one!

File: internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go
Comment (sbueringer) [inline internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go:74]: I would suggest to basically replace every occurence of helperOptions with filterOptions but let's please do it in a follow-up PR to keep the diff here smaller.

File: internal/controllers/machineset/machineset_controller.go
Comment (sbueringer) [inline internal/controllers/machineset/machineset_controller.go:374]:
```suggestion
// "metadata.annotations" from "manager" so that "capi-machineset" can own these fields and can work with SSA.
```

File: internal/controllers/machineset/machineset_controller_test.go
Comment (sbueringer) [inline internal/controllers/machineset/machineset_controller_test.go:1010]: nit: maybe just add a godoc comments here that those labels/annotations are preserved, dropped, modified later when we update a MachineSet to run an in-place mutation on Machine, BootstrapConfig and InfraMachine

(Independent of that, definitely like the new labels / annotations a lot more than what we had before)

File: internal/controllers/machineset/machineset_controller_test.go
Comment (sbueringer) [inline internal/controllers/machineset/machineset_controller_test.go:1211]:
```suggestion
	// "capi-machineset" manager.
```

File: internal/controllers/machineset/machineset_controller_test.go
Comment (sbueringer) [inline internal/controllers/machineset/machineset_controller_test.go:1124]: nit: can we change all those label/annotation validations by ones which check the exact content of the map?
```go
	g.Expect(updatedInPlaceMutatingMachine.Labels).Should(Equal(map[string]string{
	    ....	
	}))
```

I know we have to add a few additional labels, but I think it's better as it ensures that the map looks 100% exactly like we expect it too.

**Review by sbueringer — COMMENTED** [review]
Nice and very clean code, great work!

Last round of nits from my side

File: internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go
Comment (ykakarap) [reply internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go:74]: Added a follow-up item in the umbrella issue.

**Comment by sbueringer** [comment]
Great work!!

/lgtm

File: internal/util/ssa/filterintent.go
Comment (vincepri) [inline internal/util/ssa/filterintent.go:115]:
```suggestion
// IsPathAllowed returns true when the Path is one of the AllowedPaths.
func IsPathAllowed(allowedPaths []contract.Path) func(path contract.Path) bool {
```

File: internal/util/ssa/filterintent.go
Comment (vincepri) [inline internal/util/ssa/filterintent.go:130]: Should this be just the opposite of the above? Wouldn't it be better to ask users to just do `if ! <...>`

File: internal/util/ssa/filterintent.go
Comment (vincepri) [inline internal/util/ssa/filterintent.go:138]:
```suggestion
// IsPathIgnored returns true when the Path is one of the IgnorePaths.
func IsPathIgnored(ignorePaths []contract.Path) func(path contract.Path) bool {
```

File: internal/util/ssa/managedfields.go
Comment (vincepri) [inline internal/util/ssa/managedfields.go:46]:
```suggestion
func DropManagedFields(ctx context.Context, c client.Client, obj client.Object, ssaManager string, paths []contract.Path) error {
```
?

File: internal/util/ssa/filterintent.go
Comment (ykakarap) [reply internal/util/ssa/filterintent.go:115]: Since these functions where not introduced in this package and just migrated from another internal package I would like to move this discussion into a follow-up issue to unblock this PR.

Since these functions are internal we are safe to change them in a follow-up.

I added a follow-up item for this in the umbrella issue.

(Same for the other ssa util review comments).

File: internal/util/ssa/managedfields.go
Comment (ykakarap) [reply internal/util/ssa/managedfields.go]: Done.

**Review by vincepri — APPROVED** [review]

File: internal/util/ssa/filterintent.go
Comment (sbueringer) [reply internal/util/ssa/filterintent.go:115]: +1

Comment (fabriziopandini) [reply internal/util/ssa/filterintent.go:115]: +1 to fix this/similar methods in a follow-up now that those methods are part of our internal utils

File: internal/util/ssa/managedfields.go
Comment (fabriziopandini) [inline internal/util/ssa/managedfields.go:48]: q: is this function returning always true in the case of infra and boostrap machine?
can we make this smart and consider only paths? (this can be also a follow up)

**Review by fabriziopandini — COMMENTED** [review]
Great work!
/lgtm
/approve

/hold
For @ykakarap @sbueringer to evaluate my question above

File: internal/util/ssa/managedfields.go
Comment (sbueringer) [reply internal/util/ssa/managedfields.go]: @ykakarap Please do the same for CleanUpManagedFieldsForSSAAdoption either in a separate follow-up or just part of the KCP PR

File: internal/util/ssa/managedfields.go
Comment (sbueringer) [reply internal/util/ssa/managedfields.go:48]: > q: is this function returning always true in the case of infra and boostrap machine?

No.

The following happens: (example based on KubeadmConfig)
1. After create:
    * manager owns all fields including labels and annotations
2. After DropManagedFields
    * manager owns all fields **except** labels and annotations
3. After updateExternalObject:
    * manager owns all fields **except** labels and annotations
    * capi-machineset owns labels and annotations

In 2 `hasFieldsManagedBy` returned false, after 3 it returns true.

So basically this func detects correctly if we have to cleanup managed fields again by checking if capi-machineset already wrote the object. Which is exactly what we want as far as I can tell.

In which way do you want to make it smarter? That this func detects if ssaManager already owns any/all passed in paths? Checking for all could be a problem in the general case when ssaManager does not write **all** paths that have been passed in.

It's fine in the MachineSet case though as we are even getting ownership of annotations if we don't have an opinion on a specific annotation:
```yaml
          f:annotations: {}
          f:labels:
            f:cluster.x-k8s.io/cluster-name: {}
            f:cluster.x-k8s.io/deployment-name: {}
            f:cluster.x-k8s.io/set-name: {}
            f:machine-template-hash: {}
            f:topology.cluster.x-k8s.io/deployment-name: {}
            f:topology.cluster.x-k8s.io/owned: {}
```

Comment (fabriziopandini) [reply internal/util/ssa/managedfields.go:48]: thanks for the clarification

**Pull request merged**
